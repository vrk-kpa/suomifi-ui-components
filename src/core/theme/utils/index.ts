export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
