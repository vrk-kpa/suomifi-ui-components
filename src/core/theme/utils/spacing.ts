import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (props: SpacingProps | undefined): string => {
  if (!props) return '';

  const cssStyles = Object.entries(props)
    .map(([key, value]) =>
      getCSSSpacing(defaultSuomifiTheme, key as keyof SpacingProps, value),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return `margin-right: ${amount}; margin-left: ${amount};`;
    case 'my':
      return `margin-top: ${amount}; margin-bottom: ${amount};`;
    case 'px':
      return `padding-right: ${amount}; padding-left: ${amount};`;
    case 'py':
      return `padding-top: ${amount}; padding-bottom: ${amount};`;
    default:
      return `${[cssSelector[key]]}: ${amount};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
