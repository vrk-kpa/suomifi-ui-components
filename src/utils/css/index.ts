export { alphaHex, hslaToHex } from './colors';
export { allStates, absolute } from './pseudo';
export {
  clearfix,
  cssObjectToCss,
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from './utils';
