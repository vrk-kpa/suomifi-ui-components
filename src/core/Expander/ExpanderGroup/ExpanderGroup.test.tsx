import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../../utils/test';
import {
  ExpanderGroup,
  ExpanderGroupProps,
  Expander,
  ExpanderProps,
  ExpanderTitleButton,
  ExpanderTitleButtonProps,
  ExpanderContent,
} from '../';

const TestExpanderWithProps = (
  props: Omit<ExpanderProps, 'children'>,
  titleProps: ExpanderTitleButtonProps & { 'data-testid': string },
  content: string,
  key: number,
) => {
  const { children: title, ...titlePassProps } = titleProps;
  return (
    <Expander key={key} {...props}>
      <ExpanderTitleButton {...titlePassProps}>{title}</ExpanderTitleButton>
      <ExpanderContent>{content}</ExpanderContent>
    </Expander>
  );
};

const TestExpanderGroup = (
  expanderData: {
    expanderProps: Omit<ExpanderProps, 'children'>;
    titleProps: ExpanderTitleButtonProps & { 'data-testid': string };
    content: string;
  }[],
  expanderGroupProps?: Partial<ExpanderGroupProps>,
) => (
  <ExpanderGroup
    openAllText="Open all"
    closeAllText="Close all"
    ariaOpenAllText="Open all expanders"
    ariaCloseAllText="Close all expanders"
    {...expanderGroupProps}
  >
    {expanderData.map((d, index) =>
      TestExpanderWithProps(d.expanderProps, d.titleProps, d.content, index),
    )}
  </ExpanderGroup>
);

const basicExpanderProps = [
  {
    expanderProps: {
      id: 'id-first',
    },
    titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
    content: 'First content',
  },
  {
    expanderProps: {
      id: 'id-second',
    },
    titleProps: { 'data-testid': 'expander-title-2', children: 'Second' },
    content: 'Second content',
  },
];

describe('Basic expander group', () => {
  it('should not remount when calling render with the same component on the same container', () => {
    const { getByTestId, rerender } = render(
      TestExpanderGroup(basicExpanderProps),
    );
    expect(getByTestId('expander-title-1').textContent).toBe('First');
    rerender(
      TestExpanderGroup([
        {
          expanderProps: {
            id: 'id-first-2',
          },
          titleProps: {
            'data-testid': 'expander-title-1-1',
            children: 'First but not the best',
          },
          content: 'First but not the content',
        },
        {
          expanderProps: {
            id: 'id-second-2',
          },
          titleProps: { 'data-testid': 'expander-title-2', children: 'Second' },
          content: 'Second content',
        },
      ]),
    );
    expect(getByTestId('expander-title-1-1').textContent).toBe(
      'First but not the best',
    );
  });
  it('should match snapshot', () => {
    const { container } = render(TestExpanderGroup(basicExpanderProps));
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('default behaviour', () => {
  const DefaultGroup = TestExpanderGroup(
    [
      {
        expanderProps: {
          id: 'id-first',
        },
        titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
        content: 'First content',
      },
      {
        expanderProps: {
          id: 'id-second',
        },
        titleProps: {
          'data-testid': 'expander-title-2',
          children: 'Second',
          toggleButtonProps: { 'data-testid': 'expander-title-2-button' },
        },
        content: 'Second content',
      },
    ],
    { toggleAllButtonProps: { 'data-testid': 'open-all-button' } },
  );
  it('open/close all should open/close the Expanders', () => {
    const { getByTestId } = render(DefaultGroup);
    const titleDiv = getByTestId('expander-title-2');
    const button = getByTestId('expander-title-2-button');
    expect(titleDiv).not.toHaveClass('fi-expander_title-button--open');
    expect(button.querySelector('svg')).not.toHaveClass(
      'fi-expander_title-button-icon--open',
    );
    fireEvent.click(button);
    expect(titleDiv).toHaveClass('fi-expander_title-button--open');
    expect(button.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
    fireEvent.click(button);
    expect(titleDiv).not.toHaveClass('fi-expander_title-button--open');
    expect(button.querySelector('svg')).not.toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });
  it('open/close all should have providedTexts and screen reader texts', async () => {
    const { getByTestId, getByText } = render(DefaultGroup);
    const buttonVisibleText = getByText('Open all');
    expect(buttonVisibleText).toHaveAttribute('aria-hidden', 'true');
    const buttonAriaText = getByText('Open all expanders');
    expect(buttonAriaText).toHaveClass('fi-visually-hidden');
    const openAllbutton = getByTestId('open-all-button');
    fireEvent.click(openAllbutton);
    const buttonVisibleCloseText = getByText('Close all');
    expect(buttonVisibleCloseText).toHaveAttribute('aria-hidden', 'true');
    const buttonAriaCloseText = getByText('Close all expanders');
    expect(buttonAriaCloseText).toHaveClass('fi-visually-hidden');
  });
});

describe('Toggle all button', () => {
  it('should be hidden when showToggleAllButton is false', () => {
    const { queryByTestId } = render(
      TestExpanderGroup(
        [
          {
            expanderProps: {
              id: 'id-first',
            },
            titleProps: {
              'data-testid': 'expander-title-1',
              children: 'First',
            },
            content: 'First content',
          },
          {
            expanderProps: {
              id: 'id-second',

              defaultOpen: true,
            },
            titleProps: {
              'data-testid': 'expander-title-2',
              children: 'Second',
            },
            content: 'Second content',
          },
        ],
        {
          toggleAllButtonProps: {
            'data-testid': 'toggle-all-button',
          },
          showToggleAllButton: false,
        },
      ),
    );
    const toggleAllButton = queryByTestId('toggle-all-button');
    expect(toggleAllButton).toBeNull();
  });
});

describe('defaultOpen', () => {
  it('gives the classname to expander title and icon', () => {
    const { getByTestId } = render(
      TestExpanderGroup([
        {
          expanderProps: {
            id: 'id-first',
          },
          titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
          content: 'First content',
        },
        {
          expanderProps: {
            id: 'id-second',

            defaultOpen: true,
          },
          titleProps: { 'data-testid': 'expander-title-2', children: 'Second' },
          content: 'Second content',
        },
      ]),
    );
    const titleDiv = getByTestId('expander-title-2');
    expect(titleDiv).toHaveClass('fi-expander_title-button--open');
    expect(titleDiv.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });

  it('classnames will be removed when clicked', () => {
    const mockClickHandler = jest.fn();
    const { getByTestId } = render(
      TestExpanderGroup([
        {
          expanderProps: {
            id: 'id-first',
          },
          titleProps: {
            'data-testid': 'expander-title-1',
            children: 'First',
          },
          content: 'First content',
        },
        {
          expanderProps: {
            id: 'id-second',
            defaultOpen: true,
            onOpenChange: mockClickHandler,
          },
          titleProps: {
            'data-testid': 'expander-title-2',
            children: 'Second',
            toggleButtonProps: {
              'data-testid': 'expander-title-2-button',
            },
          },
          content: 'Second content',
        },
      ]),
    );
    const button = getByTestId('expander-title-2-button');
    const titleDiv = getByTestId('expander-title-2');
    expect(titleDiv).toHaveClass('fi-expander_title-button--open');
    fireEvent.click(button);
    expect(button.querySelector('svg')).not.toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });
});

describe('onOpenChange', () => {
  it('is called', () => {
    const mockClickHandler = jest.fn();
    const { getByTestId } = render(
      TestExpanderGroup([
        {
          expanderProps: {
            id: 'id-first',
          },
          titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
          content: 'First content',
        },
        {
          expanderProps: {
            onOpenChange: mockClickHandler,
          },
          titleProps: {
            'data-testid': 'expander-title-2',
            children: 'Second',
            toggleButtonProps: {
              'data-testid': 'expander-title-2-button',
            },
          },
          content: 'Second content',
        },
      ]),
    );
    const button = getByTestId('expander-title-2-button');
    fireEvent.click(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
  });
});

describe('open', () => {
  it('open-classnames should be found ', async () => {
    const { getByTestId } = render(
      TestExpanderGroup([
        {
          expanderProps: {
            id: 'id-first',
            open: true,
          },
          titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
          content: 'First content',
        },
        {
          expanderProps: {
            id: 'id-second',
            open: true,
          },
          titleProps: {
            'data-testid': 'expander-title-2',
            children: 'Second',
          },
          content: 'Second content',
        },
      ]),
    );
    const titleDiv = getByTestId('expander-title-2');
    expect(titleDiv).toHaveClass('fi-expander_title-button--open');

    expect(titleDiv.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });

  it('is clicked. Should not change as it is controlled outside', async () => {
    const mockClickHandler = jest.fn();
    const { getByTestId } = render(
      TestExpanderGroup([
        {
          expanderProps: {
            id: 'id-first',
            open: true,
          },
          titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
          content: 'First content',
        },
        {
          expanderProps: {
            id: 'id-second',
            onOpenChange: mockClickHandler,
            open: true,
          },
          titleProps: {
            'data-testid': 'expander-title-2',
            children: 'Second',
            toggleButtonProps: {
              'data-testid': 'expander-title-2-button',
            },
          },
          content: 'Second content',
        },
      ]),
    );
    const button = getByTestId('expander-title-2-button');
    const titleDiv = getByTestId('expander-title-2');
    expect(titleDiv).toHaveClass('fi-expander_title-button--open');
    expect(titleDiv.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
    fireEvent.click(button);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    expect(titleDiv).toHaveClass('fi-expander_title-button--open');
    expect(titleDiv.querySelector('svg')).toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });

  it('open/close all clicked should not force the state to change.', () => {
    const { getByTestId } = render(
      TestExpanderGroup(
        [
          {
            expanderProps: {
              id: 'id-first',
              open: false,
            },
            titleProps: {
              'data-testid': 'expander-title-1',
              children: 'First',
            },
            content: 'First content',
          },
          {
            expanderProps: {
              id: 'id-second',
              open: false,
            },
            titleProps: {
              'data-testid': 'expander-title-2',
              children: 'Second',
              toggleButtonProps: {
                'data-testid': 'expander-title-2-button',
              },
            },
            content: 'Second content',
          },
        ],
        {
          toggleAllButtonProps: {
            'data-testid': 'open-all-button',
          },
        },
      ),
    );
    const titleDiv = getByTestId('expander-title-2');
    const button = getByTestId('expander-title-2-button');
    expect(titleDiv).not.toHaveClass('fi-expander_title-button--open');
    expect(button.querySelector('svg')).not.toHaveClass(
      'fi-expander_title-button-icon--open',
    );
    const openAllButton = getByTestId('open-all-button');

    fireEvent.click(openAllButton);
    expect(titleDiv).not.toHaveClass('fi-expander_title-button--open');
    expect(button.querySelector('svg')).not.toHaveClass(
      'fi-expander_title-button-icon--open',
    );
  });
});

describe('margin prop', () => {
  it('should have margin style from margin prop', () => {
    const { getByTestId } = render(
      <ExpanderGroup
        openAllText="Open all"
        ariaOpenAllText="Open all expanders"
        closeAllText="Close all"
        ariaCloseAllText="Close all expanders"
        data-testid="expander-group"
        margin="xs"
      >
        <Expander>Test</Expander>
      </ExpanderGroup>,
    );
    const div = getByTestId('expander-group');
    expect(div).toHaveStyle('margin: 10px');
  });
});

it('should have margin style overridden by style prop', async () => {
  const { getByTestId } = render(
    <ExpanderGroup
      openAllText="Open all"
      ariaOpenAllText="Open all expanders"
      closeAllText="Close all"
      ariaCloseAllText="Close all expanders"
      data-testid="expander-group"
      margin="xs"
      style={{ margin: 2 }}
    >
      <Expander>Test</Expander>
    </ExpanderGroup>,
  );
  const div = getByTestId('expander-group');
  expect(div).toHaveAttribute('style', 'margin: 2px;');
});

test(
  'should not have basic accessibility issues',
  axeTest(
    TestExpanderGroup([
      {
        expanderProps: {
          id: 'id-first',
        },
        titleProps: { 'data-testid': 'expander-title-1', children: 'First' },
        content: 'First content',
      },
      {
        expanderProps: {
          id: 'id-second',
        },
        titleProps: { 'data-testid': 'expander-title-2', children: 'Second' },
        content: 'Second content',
      },
    ]),
  ),
);
