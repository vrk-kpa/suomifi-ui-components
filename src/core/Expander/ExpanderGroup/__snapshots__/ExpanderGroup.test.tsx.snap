// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic expander group should match snapshot 1`] = `
.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c2:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c2::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c2::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c2::-webkit-inner-spin-button {
  height: auto;
}

.c2::-webkit-outer-spin-button {
  height: auto;
}

.c2 ::before,
.c2 ::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 ::before,
.c0 ::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c5 ::before,
.c5 ::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c3 ::before,
.c3 ::after {
  box-sizing: border-box;
}

.c4 {
  position: absolute;
  clip: rect(0 0 0 0);
  height: 1px;
  width: 1px;
  margin: -1px;
  padding: 0;
  border: 0;
  overflow: hidden;
}

.c1 {
  color: hsl(0, 0%, 13%);
  width: 100%;
  max-width: 100%;
}

.c1.fi-expander-group {
  display: flex;
  flex-direction: column;
}

.c1>.fi-expander-group_expanders {
  flex: none;
}

.c1>.fi-expander-group_expanders .fi-expander {
  margin-top: 0;
  margin-bottom: 0;
  border-radius: 0;
  transition: margin 50ms cubic-bezier(0.28, 0.84, 0.42, 1);
}

.c1>.fi-expander-group_expanders .fi-expander> {
  border-radius: 0;
}

.c1>.fi-expander-group_expanders .fi-expander:first-child {
  border-radius: 2px 2px 0 0;
}

.c1>.fi-expander-group_expanders .fi-expander:last-child {
  border-bottom-left-radius: 2px;
  border-bottom-right-radius: 2px;
}

.c1>.fi-expander-group_expanders .fi-expander:not(:first-child) {
  border-top: none;
}

.c1>.fi-expander-group_expanders .fi-expander.fi-expander--open {
  border-top: 1px solid hsl(212, 63%, 45%);
  border-radius: 2px;
}

.c1>.fi-expander-group_expanders .fi-expander.fi-expander--open+.fi-expander {
  border-top: 1px solid hsl(212, 63%, 45%);
}

.c1>.fi-expander-group_expanders .fi-expander.fi-expander--open:not(:first-of-type) {
  margin-top: 16px;
}

.c1>.fi-expander-group_expanders .fi-expander.fi-expander--open:not(:last-of-type) {
  margin-bottom: 16px;
}

.c1>.fi-expander-group_all-button {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  flex: 1 1 auto;
  align-self: flex-end;
  margin-left: auto;
  margin-bottom: 8px;
  padding: 4px 0;
  color: hsl(212, 63%, 45%);
  border-radius: 2px;
  border: none;
  cursor: pointer;
}

.c1>.fi-expander-group_all-button:focus {
  outline: 3px solid transparent;
  position: relative;
}

.c1>.fi-expander-group_all-button:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0, 0%, 100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  z-index: 9999;
}

.c1>.fi-expander-group_all-button :not(:focus-visible) {
  outline: 0;
}

.c1>.fi-expander-group_all-button :not(:focus-visible):after {
  content: none;
}

.c6 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0, 0%, 100%);
  position: relative;
  padding: 0;
  border: 1px solid hsl(212, 63%, 45%);
  border-radius: 2px;
  width: 100%;
  max-width: 100%;
}

.c6.fi-expander {
  display: block;
}

.c6:before {
  background-color: hsl(212, 63%, 98%);
  opacity: 0;
}

.c6.fi-expander--open {
  border-bottom: none;
}

.c9 {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  background-color: hsl(0, 0%, 100%);
  border-radius: inherit;
  position: relative;
  visibility: hidden;
  height: 0;
  overflow: hidden;
  word-break: break-word;
  transform-origin: top;
}

.c9.fi-expander {
  display: block;
}

.c9:not(.fi-expander_content--no-padding) {
  padding: 0 16px;
}

.c9.fi-expander_content--open {
  visibility: visible;
  height: auto;
  border-bottom: 1px solid hsl(212, 63%, 45%);
  border-top: 1px solid transparent;
  border-top-left-radius: 0;
  border-top-right-radius: 0;
}

.c9.fi-expander_content--open:not(.fi-expander_content--no-padding) {
  padding: 10px 20px 20px;
}

.c8 {
  vertical-align: baseline;
}

.c8.fi-icon {
  display: inline-block;
}

.c8 .fi-icon-base-fill {
  fill: currentColor;
}

.c8 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c8.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c8.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c7 {
  color: hsl(0, 0%, 13%);
  position: relative;
  width: 100%;
  max-width: 100%;
  min-height: 60px;
  background-color: hsl(212, 63%, 98%);
  border-radius: inherit;
}

.c7:hover {
  background: linear-gradient(0deg, hsl(215, 100%, 95%), hsl(215, 100%, 97%));
}

.c7 .fi-expander_title-button {
  display: block;
}

.c7.fi-expander_title-button--open {
  background-color: hsl(0, 0%, 100%);
  border-bottom-left-radius: 0;
  border-bottom-right-radius: 0;
}

.c7 .fi-expander_title-button_button {
  color: hsl(0, 0%, 13%);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: calc(20 / 14);
  letter-spacing: 0;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(212, 63%, 45%);
  display: inline-block;
  width: 100%;
  max-width: 100%;
  min-height: 60px;
  padding: 17px 60px 16px 20px;
}

.c7 .fi-expander_title-button_button:focus {
  outline: 0;
}

.c7 .fi-expander_title-button_button:focus-within {
  outline: 3px solid transparent;
}

.c7 .fi-expander_title-button_button:focus-within:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0, 0%, 100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196, 77%, 44%);
  z-index: 9999;
}

.c7 .fi-expander_title-button_button :not(:focus-visible) {
  outline: 0;
}

.c7 .fi-expander_title-button_button :not(:focus-visible):after {
  content: none;
}

.c7 .fi-expander_title-button_button * {
  cursor: pointer;
}

.c7 .fi-expander_title-button_button:hover,
.c7 .fi-expander_title-button_button:active,
.c7 .fi-expander_title-button_button:focus,
.c7 .fi-expander_title-button_button:focus-within {
  cursor: pointer;
}

.c7 .fi-expander_title-button-icon {
  position: absolute;
  top: 0;
  right: 0;
  height: 20px;
  width: 20px;
  margin: 20px;
}

.c7 .fi-expander_title-button--open .fi-expander_title-button-icon,
.c7 .fi-expander_title-button-icon--open {
  transform: rotate(-180deg);
}

<div
  class="c0 c1 fi-expander-group"
>
  <button
    class="c2 fi-expander-group_all-button"
    type="button"
  >
    <span
      aria-hidden="true"
      class="c3"
    >
      Open all
    </span>
    <span
      class="c3 c4 fi-visually-hidden"
    >
      Open all expanders
    </span>
  </button>
  <div
    class="c0 fi-expander-group_expanders"
  >
    <div
      class="c5 c6 fi-expander"
      id="id-first"
    >
      <div
        class="c0 c7 fi-expander_title-button"
        data-testid="expander-title-1"
      >
        <span
          class="c3"
        >
          <button
            aria-controls="id-first_content"
            aria-expanded="false"
            class="c2 fi-expander_title-button_button"
            type="button"
          >
            <span
              class="c3"
              id="id-first_title"
            >
              First
            </span>
            <svg
              aria-hidden="true"
              class="fi-icon c8 fi-expander_title-button-icon"
              focusable="false"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fi-icon-base-fill"
                d="M12.99 15.61c-.273.26-.632.39-.99.39s-.717-.13-.99-.39l-5.6-5.334a1.287 1.287 0 0 1 0-1.885 1.448 1.448 0 0 1 1.98 0l4.61 4.39 4.61-4.39a1.448 1.448 0 0 1 1.98 0c.547.52.547 1.365 0 1.885l-5.6 5.333Z"
                fill="#222"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </span>
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="id-first_title"
        class="c0 c9 fi-expander_content"
        id="id-first_content"
        role="region"
      >
        First content
      </div>
    </div>
    <div
      class="c5 c6 fi-expander"
      id="id-second"
    >
      <div
        class="c0 c7 fi-expander_title-button"
        data-testid="expander-title-2"
      >
        <span
          class="c3"
        >
          <button
            aria-controls="id-second_content"
            aria-expanded="false"
            class="c2 fi-expander_title-button_button"
            type="button"
          >
            <span
              class="c3"
              id="id-second_title"
            >
              Second
            </span>
            <svg
              aria-hidden="true"
              class="fi-icon c8 fi-expander_title-button-icon"
              focusable="false"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fi-icon-base-fill"
                d="M12.99 15.61c-.273.26-.632.39-.99.39s-.717-.13-.99-.39l-5.6-5.334a1.287 1.287 0 0 1 0-1.885 1.448 1.448 0 0 1 1.98 0l4.61 4.39 4.61-4.39a1.448 1.448 0 0 1 1.98 0c.547.52.547 1.365 0 1.885l-5.6 5.333Z"
                fill="#222"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </span>
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="id-second_title"
        class="c0 c9 fi-expander_content"
        id="id-second_content"
        role="region"
      >
        Second content
      </div>
    </div>
  </div>
</div>
`;
